Add support for Gardener pre-release integration test pipeline

MrBatschner opened the pull request:
**What this PR does / why we need it**:

This PR adds support for a Gardener integration test pipeline that supports testing Garden Linux versions that have not been release (yet). The changes include a dedicated publishing config and several changes that allow for publishing temporary artefcats to non-productive locations.

Supports AWS only at the moment, the other platforms will follow in seperate PRs.

**Special notes for your reviewer**:

You might want to review the individual commits separately and have a look at their commit messages to understand their intentions.

This PR should not be squashed, the commits should be left intact upon merging.

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```feature operator
The glci publishing gear now supports publishing and cleaning of temporary artefacts that can be used in Gardener integration tests.
```

MrBatschner commented:
/ready
/invite @ccwienk 
/invide @zkdev

ccwienk commented on glci/alicloud.py:
```diff
@@ -265,7 +266,13 @@ def _list_regions(self) -> []:
+            if region_id == self.region:
+                continue
+
+            if self.regions_to_copy_to != None:
```
style: according to pep8, you should use identity-comparison for singletons such as `None` (is instead of == / !=)

however, this might even be improved:

```
# single if
if self.regions_to_copy_to and region_id in self.regions_to_copy_to:
  ...

# or: initialise region_to_copy_to as empty collection
if region in self.regions_to_copy_to:
 ...
```

ccwienk commented on glci/aws.py:
```diff
@@ -179,9 +179,15 @@ def register_image(
     for region in ec2_client.describe_regions()['Regions']:
-        yield region['RegionName']
+        region_name = region['RegionName']
+        if regions_to_include != None:
```
style: use `is` | `is not` (or rather: omit comparison to None)

ccwienk commented on glci/model.py:
```diff
@@ -715,13 +728,15 @@ class PublishingTargetAliyun:
     aliyun_cfg_name: str
     oss_bucket_name: str
     aliyun_region: str
+    copy_regions: typing.Optional[list[str]]
```
nit: if you initialise to empty list | tuple you might omit checking for None (-> less special-cases)

ccwienk commented on .cicd-cli.py:
```diff
@@ -500,14 +506,21 @@ def ls_manifests():
         '--print',
         default='all',
-        choices=('all', 'versions', 'versions-and-commits'),
+        choices=('all', 'versions', 'versions-and-commits', 'latest'),
```
considering you sort by version, this should rather be called `greatest`

ccwienk commented on .cicd-cli.py:
```diff
@@ -525,31 +539,49 @@ def iter_manifest_prefixes():
+            )
+            manifests.append(s)
+
+    manifests.sort(key=lambda v: float(v.version))
```
rather use cc-util's `version.parse_to_semver` (parses to "relaxed semver"). or: split into two integers, and implement custom comparison. said module btw also features `version.greatest_version`. if you would prefer to not introduce this dependency, feel free to copy-paste required parts from version.py

MrBatschner replied:
Actually, I am not entirely sure what I was thinking when I was using a `float` for it...

ccwienk commented on .cicd-cli.py:
```diff
@@ -525,31 +539,49 @@ def iter_manifest_prefixes():
+
+    manifests.sort(key=lambda v: float(v.version))
+
+    if parsed.print == 'latest':
```
greatest

ccwienk commented on .cicd-cli.py:
```diff
@@ -525,31 +539,49 @@ def iter_manifest_prefixes():
+                committish=m.committish
+            )
+            with open(parsed.yaml[0], "w") as f:
+                f.write(yaml.dump(v.__dict__))
```
safe_dump? also, preferably use `dataclasses.asdict`

ccwienk commented on .cicd-cli.py:
```diff
@@ -651,27 +651,41 @@ def publish_release_set():
-    if not bool(parsed.commit) ^ bool(parsed.version_name):
-        logger.fatal('exactly one of --commit, --version-name must be passed')
-        exit(1)
+    if not bool(parsed.infile):
```
for convenience, you might also check for `sys.stdin.isatty()` and read from stdin in this case

ccwienk commented on .cicd-cli.py:
```diff
@@ -442,6 +440,8 @@ def _publishing_cfg(parsed):
 
 
 def replicate_blobs():
+    import replicate    # late import because unneeded for the other functions
```
hmm.... I guess late-imports are okay-ish (not a nice codestyle). is this motivated by significant performance costs from those imports?

MrBatschner replied:
These modules (`replicate`, `publish`, `cleanup`) will eventually import lots of the hyperscaler SDKs. For something simple as `ls_manifests`, which only requires boto3, the presence of Azure or AliCloud SDK modules should not be a hard requirement - hence the late imports. I guess this is one of the tradeoffs to be considered when using the busybox swiss-army knife approach...

ccwienk commented on .cicd-cli.py:
```diff
@@ -910,6 +915,160 @@ def end_phase(name):
+        '--version',
+    )
+    parser.add_argument(
+        '--ctx-repo',
```
nit: since rename from "CNUDIE" -> OCM, those are actually called "OCM Repository" / -ies (so --ocm-repo might be nicer)

ccwienk commented on .cicd-cli.py:
```diff
@@ -910,6 +915,160 @@ def end_phase(name):
+    parsed = parser.parse_args()
+
+    if bool(parsed.infile):
+        with open(parsed.infile[0]) as f:
```
using YAML is fine. however: would a plain text file only containing version be easier for local usage?

MrBatschner replied:
True. Somehow, I wanted to have this future proof already in case the committish is required in addition to the version and in that case, I would prefer a structured file-format over a plain text file.

ccwienk commented on .cicd-cli.py:
```diff
@@ -910,6 +915,160 @@ def end_phase(name):
+    elif bool(parsed.version):
+        version = parsed.version
+    else:
+        raise RuntimeError(f"need to provide either --version or --infile parameter")
```
nit: why not call it `--version-file`?

ccwienk commented on component_descriptor.py:
```diff
@@ -106,6 +108,24 @@ def component_descriptor(
 
 
+def retrieve_component(
+        ctx_repository: str,
```
not sure whether this function is actually useful (that's what component-descriptor-lookups are for..)

rather define a helperfunction that will create and return a component-lookup.

ccwienk commented on component_descriptor.py:
```diff
@@ -106,6 +108,24 @@ def component_descriptor(
+        component_version: str,
+):
+    lookup = cnudie.retrieve.create_default_component_descriptor_lookup(
+        default_ctx_repo=cm.OciRepositoryContext(baseUrl=ctx_repository),
```
please no longer use default_ctx_repo. this is deprecated (yes, it is not yet marked as such).

instead, pass-in an ocm-repository-lookup:

```
ocm_repo_lookup = cnudie.retrieve.ocm_repository_lookup(ocm_repository_url)

component_descriptor_lookup = cnudie.retrieve.create_default_component_descriptor_lookup(
 ocm_repository_lookup=ocm_repo_lookup,
)
```

`component_descriptor_lookup` btw offers some convenience for doing lookups:

```
# use single str
comp = lookup('github.com/gardenlinux/gardenlinux:1300.0')+
# use two-tuple (name, version)
comp = lookup(('github.com/gardenlinux/gardenlinux', '1300.0')
# ... also allowed: Component, ComponentDescriptor, ComponentIdentity, ComponentReference

ccwienk commented on glci/aws.py:
```diff
@@ -317,13 +318,82 @@ def unregister_images_by_name(
+        Owners=["self"], Filters=[{"Name": "image-id", "Values": [image_id]}]
+    )
+
+    if len(response["Images"]) > 0:
```
style: need not check for leng (if response['Images']: ...)

ccwienk commented on glci/aws.py:
```diff
@@ -317,13 +318,82 @@ def unregister_images_by_name(
+
+    if len(snapshots) < 1:
+        logger.warning(f"no snapshots found for {image_id=}")
+    return snapshots
```
style: might use a generator (yield)

MrBatschner replied:
If the generator is passed into the `_delete_snapshots()` (I guess the most meaningful way of using a generator),  `ec2.describe_images()` to obtain the snapshot IDs would be called AFTER `ec2.deregister_image()` which - depending how fast AWS is with actually deregistering the image - could lead to an empty response and dangling snapshots.

ccwienk commented on glci/aws.py:
```diff
@@ -317,13 +318,82 @@ def unregister_images_by_name(
+
+
+def _delete_snapshots(
+        client: 'botocore.client.EC2',
```
style: instead of passing-around dry_run-flags, you might pass-in dummy-clients (might not be a good idea overall, though)

MrBatschner replied:
Brilliant idea, something I will probably implement in a new, separate PR.

ccwienk replied:
yeah, it will have limitations if responses of such dummy-clients are required. although that might of course also be dealt somehow.

ccwienk commented on glci/aws.py:
```diff
@@ -317,13 +318,82 @@ def unregister_images_by_name(
+    images: glci.model.AwsPublishedImageSet,
+    dry_run: bool
+):
+    executor = concurrent.futures.ThreadPoolExecutor(max_workers=len(images))
```
TPE will default to max_workers equal number of logical CPUs. this might be a better choice than number of images. unless you do not care about number of vCPUs and really one a dedicated thread per image

ccwienk commented on glci/aws.py:
```diff
@@ -443,6 +513,7 @@ def upload_and_register_gardenlinux_image(
             unregister_images_by_name(
                 mk_session=mk_session,
                 image_name=target_image_name,
+                dry_run=False,
```
should this be made a parameter that can also be passed to calling function?

ccwienk commented on cleanup.py:
```diff
@@ -1,28 +1,186 @@
+    )
+
+
+## functions moved over from clean.py
```
I think this comment is not helpful (or actually confusing) 🤔

AndreasBurger commented on .cicd-cli.py:
```diff
@@ -890,7 +890,12 @@ def end_phase(name):
     if parsed.print_component_descriptor:
         pprint.pprint(component_descriptor)
 
+    repository_context = component_descriptor.component.repositoryContexts[0].baseUrl
```
consider using current_repository_context (https://github.com/gardener/cc-utils/blob/a679e13c4a6fce91e0301ae407ab513466657414/gci/componentmodel.py#L517)

MrBatschner commented on .cicd-cli.py:
```diff
@@ -698,15 +698,19 @@ def end_phase(name):
-    s3_session = ccc.aws.session(cfg.origin_buildresult_bucket.aws_cfg_name)
+    source_manifest_bucket = cfg.source_manifest_bucket
+    target_manifest_buckets = tuple(cfg.target_manifest_buckets)
+    if len(target_manifest_buckets) == 0:
```
The reason for the - at first unintuitive - falling back to source manifest buckets is this (see line 861):
- In the normal case (publishing for productive usage), we have one release manifest which lives in the `source_manifest_bucket`. We publish the artefacts and then update the release manifest in the `source_manifest_bucket`. This is how it has been before this PR and how it should remain for productive publishing.
- In the integration test case (only temporary publishing for an integration test with timely cleanup/deletion of the published artefacts after), we have a copy of the release manifest containing the publish information just for the artefacts published for testing. That copy goes into a `target_manifest_bucket`, leaving the original manifest in the `source_manifest_bucket` untouched - the original manifest in the `source_manifest_bucket` should only contain publish information about artefacts that were released for productive usage.
- If no `target_manifest_bucket` is specified in the `publishing-cfg`, no copy of the release-manifest should be created. Instead, the original release manifest (the one in the `source_manifest_bucket`) should be amended with the publish-information.
- Thus, when it comes to updating the release manifest, if no dedicated `target_manifest_bucket` is specified, the `source_manifest_bucket` also becomes the `target_manifest_bucket` for the updated release manifest.

MrBatschner commented:
@ccwienk, @AndreasBurger : I addressed your feedback, PTAL. I will eventually squash the feedback commits before merging but wanted to keep them separate for now to simplify the review for you.
